sparrowpy/dsp.py

```diff
@@ -340,3 +339,146 @@ def energy_time_curve_from_impulse_response(
             signal.time[..., lower:upper]**2, axis=-1) * bw_factor
 
     return etc
+
+def band_filter_signal(signal:pf.Signal,
+                       frequencies:np.ndarray,
+                       num_fractions:int,
+                       order=4):
+    r"""Band filter input signal and return corresponding bandwidths.
+
+    This method filters an input signal into octave/3rd octave bands
+    with Butterworth filtering. By default, the filter order is set
+    as 4.
+    The user may input an arbitrary array of frequencies in
+    which to filter the signal. The closest fractional octave bands
+    to the input frequencies will be returned in the same order.
+
+    Returns the band-filtered signal in individual signal channels.
+
+    Parameters
+    ----------
+    signal: :py:class:`pyfar.Signal`
+        Input broad-spectrum signal.
+    frequencies: np.ndarray
+        Frequencies in which to filter the signal. In Hz.
+    num_fractions: int
+        The number of bands an octave is divided into. E.g., ``1`` refers to
+        octave bands and ``3`` to third octave bands.
+        All positive integers are allowed.
+    order: int
+        Butterworth filter order for the signal filtering. By default 4.
+
+    Returns
+    -------
+    band_filtered_noise: :py:class:`pyfar.Signal`
+        Band-filtered signal of cshape (signal.cshape + (len(freqs))
+    bandwidth: np.ndarray
+        array of bandwidth values in Hz of the fractional octave bands
+        corresponding to the input frequencies.
+    """
+    if (frequencies<=0).any():
+        raise ValueError(
+            "Input frequencies must be greater than zero.",
+        )
+    if num_fractions<=0:
+        raise ValueError(
+            "Number of octave fractions must be greater than zero.")
+
+    frequency_range = (.67*np.min(frequencies),1.5*np.max(frequencies))
+
+    bandwidth,idcs = _closest_frac_octave_data(frequencies=frequencies,
+                                         num_fractions=num_fractions,
+                                         frequency_range=frequency_range)
+
+    band_filtered_noise = pf.dsp.filter.fractional_octave_bands(
+        signal=signal,
+        num_fractions=num_fractions,
+        order=order,
+        frequency_range=frequency_range,
+    )
+
+    band_filtered_noise.time = np.squeeze(band_filtered_noise.time[idcs])
```

Comment:
```suggestion
    band_filtered_noise.time = band_filtered_noise.time[idcs]
```

why do we squeeze? I would remove it, it will lead to unexpected behaviour.

Reply:
pf.dsp.filter._fractional_octave_bands() seems to return a (1,) shaped signal instead of (1,1) when the frequency range is less than a single frequency band.
The squeezing made sure that if the input signal shape was (1,), the output has the shape (n_bands,).

But you're right that squeezing is a bad principle. 
Instead, I will force the function output to return (n_signal_channels,n_bands) every time by applying a np.newaxis in the exception case (returning a signal of shape (1,1).